### PR TITLE
✨ Add reservednames admission plugin to surface clearer error messages

### PR DESCRIPTION
**Pull request by hasheddan**
## Summary

Adds a mutating admission plugin for ClusterWorkspaces and
ClusterWorkspaceTypes to ensure that a more helfpul error message is
returned if a user attempts to create with a reserved name. A mutating
webhook is used due to the fact that we want to catch this prior to
OpenAPI schema validation.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Before
```
🤖 (kcp) k kcp ws create root
error: ClusterWorkspace.tenancy.kcp.dev "root" is invalid: <nil>: Invalid value: "": "metadata.name" must not validate the schema (not)
exit status 1
```
After
```
🤖 (kcp) k kcp ws create root
error: clusterworkspaces.tenancy.kcp.dev "root" is forbidden: cannot use reserved name "root" as "metadata.name"
exit status 1
```

## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/issues/1214

### CODE REVIEW

File: pkg/apis/tenancy/v1alpha1/helper/helper.go [hunk 1]
```diff
@@ -40,3 +40,12 @@ func QualifiedObjectName(obj metav1.Object) string {
...
 	return fmt.Sprintf("%s|%s", logicalcluster.From(obj), obj.GetName())
 }
+
+// IsReservedName indicates that the provided name is reserved and cannot be
+// used in a user-supplied object.
+func IsReservedName(name string) bool {
+	if name == "root" || name == "system" {
+		return true
+	}
+	return false
```
Comment (ncdc) [inline pkg/apis/tenancy/v1alpha1/helper/helper.go:50]: simplification suggestion: `return name == "root" || name == "system"`

Comment (ncdc) [reply pkg/apis/tenancy/v1alpha1/helper/helper.go:50]: Also: reserved clusterworkspace names are `root` and `system`. Reserved clusterworkspacetype names are `system` and `any`. (see the 2 `.yaml-patch` files in `config/crds` for the respective crds).

File: pkg/admission/reservednames/admission_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		},
+		"NotApplicableKind": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "NotaCWorCWT", "clusterworkspaces"),
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			plugin := &reservedNames{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+				applicableKinds: map[schema.GroupKind]struct{}{
```
Comment (ncdc) [inline pkg/admission/reservednames/admission_test.go:89]: Maybe we should make a constructor for the `reservedNames` instance and call that here?

**Review by hasheddan — COMMENTED** [review]
@ncdc I did a bit of a refactor here to try to make the checks a bit cleaner. Also added some `TODOs` for generating the patches we apply to the OpenAPI schema from the reserved name definitions in code. Happy to open an issue tracking that work if you're on board.

File: pkg/admission/reservednames/admission.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// Ensure that the required admission interfaces are implemented.
+// NOTE(hasheddan): we must use mutation rather than validation because OpenAPI
+// validation will occur prior to validation webhook.
+var _ = admission.MutationInterface(&ReservedNames{})
+
+// Admit ensures that the object does not violate reserved name constraints.
+func (o *ReservedNames) Admit(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	resource, kind := a.GetResource().GroupResource(), a.GetKind().GroupKind()
+	for _, fn := range o.reservedNameFns {
+		if fn(resource, kind, a.GetName()) {
```
Comment (ncdc) [inline pkg/admission/reservednames/admission.go:100]: Maybe `if isReservedName()` would be more indicative of what it's doing?

File: pkg/admission/reservednames/admission_test.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "ClusterWorkspace", "notacworcwt"),
+		},
+		"NotApplicableKind": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "NotaCWorCWT", "clusterworkspaces"),
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			plugin := NewReservedNames(
+				newReservedNameFn(
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply pkg/admission/reservednames/admission_test.go:87]: I could go either way -- I think I may be optimizing a bit too heavily for future use cases of reserved names, whereas we really only have these two APIs where it applies right now. I'll switch to hard-coded 👍🏻

File: pkg/admission/reservednames/admission.go (hunk 3 above)
Comment (hasheddan) [reply pkg/admission/reservednames/admission.go:100]: will do 👍🏻

File: pkg/admission/reservednames/admission_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hasheddan) [reply pkg/admission/reservednames/admission_test.go:87]: done 👍🏻

File: pkg/admission/reservednames/admission.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+// ReservedNames is an admission plugin for checking reserved object names.
+type ReservedNames struct {
+	*admission.Handler
+
+	reservedNameFns []reservedNameFn
+}
+
+// NewReservedNames constructs a new ReservedNames admission plugin.
+func NewReservedNames(fns ...reservedNameFn) *ReservedNames {
```
Comment (ncdc) [inline pkg/admission/reservednames/admission.go:74]:
```suggestion
func NewReservedNames() *ReservedNames {
```

File: pkg/admission/reservednames/admission.go (hunk 3 above)
Comment (hasheddan) [reply pkg/admission/reservednames/admission.go]: Went with `isReserved` as name is being passed

File: pkg/admission/reservednames/admission.go (hunk 5 above)
Comment (hasheddan) [reply pkg/admission/reservednames/admission.go:74]: whoops -- sorry about that, moving a little too fast

File: pkg/admission/reservednames/admission_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/admission/reservednames/admission_test.go:87]: Remove from here?

**Review by ncdc — APPROVED** [review]

**Pull request merged**